workermanager.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/taskcluster/generic-worker/gwconfig"
+	"github.com/taskcluster/taskcluster-client-go/tcworkermanager"
+)
+
+type WorkerManagerUserData struct {
+	WorkerPoolID string          `json:"workerPoolId"`
+	ProviderID   string          `json:"providerId"`
+	WorkerGroup  string          `json:"workerGroup"`
+	RootURL      string          `json:"rootUrl"`
+	WorkerConfig BootstrapConfig `json:"workerConfig"`
+}
+
+func (userData *WorkerManagerUserData) updateConfig(c *gwconfig.Config, providerType interface{}) error {
+	wp := strings.SplitN(userData.WorkerPoolID, "/", -1)
+	if len(wp) != 2 {
+		return fmt.Errorf("Was expecting WorkerPoolID to have syntax <provisionerId>/<workerType> but was %q", userData.WorkerPoolID)
+	}
+
+	c.ProvisionerID = wp[0]
+	c.WorkerType = wp[1]
+	c.WorkerGroup = userData.WorkerGroup
+	c.RootURL = userData.RootURL
+
+	// We need a worker manager client for fetching taskcluster credentials.
+	// Ensure auth is disabled in client, since we don't have credentials yet.
+	wm := c.WorkerManager()
+	wm.Authenticate = false
+	wm.Credentials = nil
+
+	workerIdentityProof, err := json.Marshal(providerType)
+	if err != nil {
+		return fmt.Errorf("Could not marshal provider type %#v: %v", providerType, err)
+	}
+
+	reg, err := wm.RegisterWorker(&tcworkermanager.RegisterWorkerRequest{
+		WorkerPoolID:        userData.WorkerPoolID,
+		ProviderID:          userData.ProviderID,
+		WorkerGroup:         userData.WorkerGroup,
+		WorkerID:            c.WorkerID,
+		WorkerIdentityProof: json.RawMessage(workerIdentityProof),
+	})
+
+	if err != nil {
+		return fmt.Errorf("Could not register worker: %v", err)
+	}
+
+	c.AccessToken = reg.Credentials.AccessToken
+	c.Certificate = reg.Credentials.Certificate
+	c.ClientID = reg.Credentials.ClientID
+
+	// TODO: process reg.Expires
```

Comment:
Should we do this todo before this lands?

Reply:
I don't see this as critical. Workers don't live too long, and if it used bad creds, it would get an error and shut itself down, and if it was running a task, resolve it as internal-error, and the task would get rescheduled for another worker, so it isn't critical, and also should only happen rarely for very long-lived workers. Workers typically shut down after 15 mins if they don't have a job, so quite unlikely.